src/BaseExtractor.php

```diff
@@ -0,0 +1,319 @@
+        $sshParameters->setPrivateKey($privateKey);
+        $sshParameters->setRemoteHost($dbConfig->getHost());
+        $sshParameters->setRemotePort($dbConfig->getPort());
+
+        if (!$sshParameters->getUser()) {
+            $sshParameters->setUser($dbConfig->getUser());
+        }
+
+        $tunnelParams = $sshParameters->toArray();
+        $this->getLogger()->info("Creating SSH tunnel to '" . $tunnelParams['sshHost'] . "'");
+        $proxy = new RetryProxy(
+            $this->getLogger(),
+            RetryProxy::DEFAULT_MAX_TRIES,
+            RetryProxy::DEFAULT_BACKOFF_INTERVAL,
+            ['SSHException', 'Exception']
+        );
+        try {
+            $proxy->call(function () use ($tunnelParams):void {
+                $ssh = new SSH();
+                $ssh->openTunnel($tunnelParams);
+            });
+        } catch (SSHException $e) {
+            throw new UserException($e->getMessage(), 0, $e);
+        }
+
+        $dbConfig->setHost('127.0.0.1');
+        $dbConfig->setPort($sshParameters->getLocalPort());
+        return $dbConfig;
+    }
+
+    protected function getConfigDefinitionClass(): string
+    {
+        $configRaw = $this->getRawConfig();
+        $action = $configRaw['action'];
+        $isConfigRow = !isset($configRaw['parameters']['tables']);
+
+        if ($action !== 'run') {
+            return ActionConfigDefinition::class;
+        } elseif ($isConfigRow) {
+            return ConfigRowDefinition::class;
+        } else {
+            return ConfigDefinition::class;
+        }
+    }
+
+    protected function getOutputFileName(string $outputTableName): string
+    {
+        $sanitizedTableName = Strings::webalize($outputTableName, '._');
+        return $sanitizedTableName . '.csv';
+    }
+
+    protected function getOutputFilePath(string $outputTableName): string
+    {
+        return $this->getDataDir() . '/out/tables/' . $this->getOutputFileName($outputTableName);
+    }
+
+    protected function checkConnectionIsAlive(): void
+    {
+        try {
+            $this->testConnection();
+        } catch (\Throwable $e) {
+            throw new DeadConnectionException("Dead connection: " . $e->getMessage());
+        }
+    }
+
+    protected function validateIncrementalFetching(
+        TableDetailParameters $table,
+        string $columnName,
+        ?int $limit = null
+    ): void {
+        throw new UserException('Incremental Fetching is not supported by this extractor.');
+    }
+
+    protected function quote(string $obj): string
+    {
+        return "\"{$obj}\"";
+    }
+
+    protected function getConfigClass(): string
+    {
+        return BaseExtractorConfig::class;
+    }
+
+    public static function getColumnMetadata(array $column): array
+    {
+        $datatype = new GenericStorage(
+            $column['type'],
+            array_intersect_key($column, array_flip(self::DATATYPE_KEYS))
+        );
+        $columnMetadata = $datatype->toMetadata();
```

Comment:
This idea is good, but for now I'm missing interface for `MssqlDataType` and `GenericStorage`.

Reply:
I think it can be simply like
```
protected function getDatatypeMetadata(array $column): array
{
     return new GenericStorage(
            $column['type'],
            array_intersect_key($column, array_flip(self::DATATYPE_KEYS))
        )->toMetadata();
}
```

Reply:
and the mssql-ex will override it and use MssqlDataType instead of GenericStorage

Reply:
Yes you're right. I was thinking about extracting whole method with only specified DataType class in each extractor.